modules/nf-core/cellranger/multi/templates/cellranger_multi.py

```diff
@@ -0,0 +1,187 @@
+    r2.rename(fastq_all / subdir / f"{sample_id}_S1_L{i:03d}_R2_001.fastq.gz")
+
+#
+# fix relative paths from main.nf
+#
+work_dir = str(Path.cwd()) + '/'
+gex_reference_path = "${gex_reference_path}".replace('./', work_dir)
+frna_probeset = "${frna_probeset}".replace('./', work_dir)
+target_panel = "${target_panel}".replace('./', work_dir)
+cmo_reference_path = "${cmo_reference_path}".replace('./', work_dir)
+cmo_barcode_path = "${cmo_barcode_path}".replace('./', work_dir)
+fb_reference_path = "${fb_reference_path}".replace('./', work_dir)
+vdj_reference_path = "${vdj_reference_path}".replace('./', work_dir)
+primer_index = "${primer_index}".replace('./', work_dir)
+fastq_gex = "${fastq_gex}".replace('./', work_dir)
+fastq_vdj = "${fastq_vdj}".replace('./', work_dir)
+fastq_antibody = "${fastq_antibody}".replace('./', work_dir)
+fastq_beam = "${fastq_beam}".replace('./', work_dir)
+fastq_crispr = "${fastq_crispr}".replace('./', work_dir)
+fastq_cmo = "${fastq_cmo}".replace('./', work_dir)
+
+#
+# generate config file for cellranger multi
+#
+config_txt = f"""${include_gex}
+{gex_reference_path}
+{frna_probeset}
+${gex_options_filter_probes}
+${gex_options_r1_length}
+${gex_options_r2_length}
+${gex_options_chemistry}
+${gex_options_expect_cells}
+${gex_options_force_cells}
+${gex_options_no_secondary}
+${gex_options_no_bam}
+${gex_options_check_library_compatibility}
+{target_panel}
+${gex_options_no_target_umi_filter}
+${gex_options_include_introns}
+${cmo_options_min_assignment_confidence}
+{cmo_reference_path}
+{cmo_barcode_path}
+
+${include_fb}
+{fb_reference_path}
+${fb_options_r1_length}
+${fb_options_r2_length}
+
+${include_vdj}
+{vdj_reference_path}
+{primer_index}
+${vdj_options_r1_length}
+${vdj_options_r2_length}
+
+[libraries]
+fastq_id,fastqs,lanes,feature_types
+{fastq_gex}
+{fastq_vdj}
+{fastq_antibody}
+{fastq_beam}
+{fastq_crispr}
+{fastq_cmo}
+"""
+
+with open("${config}", 'w') as file:
+    file.write(config_txt)
+
+#
+# check the extra data that is included
+#
+if len("${include_cmo}") > 0:
+    with open("${config}", 'a') as file, open("${cmo_barcodes}", 'r') as input_conf:
+        file.write("${include_cmo}\\n")
+        for line in input_conf:
+            file.write(line + "\\n")
+
+if len("${include_beam}") > 0:
+    with open("${config}", 'a') as file, open("${beam_csv_text}", 'r') as input_conf:
+        file.write("${include_beam}\\n")
+        for line in input_conf:
+            file.write(line + "\\n")
+
+if len("${include_frna}") > 0:
+    with open("${config}", 'a') as file, open("${frna_csv_text}", 'r') as input_conf:
+        file.write("${include_frna}\\n")
+        for line in input_conf:
+            file.write(line + "\\n")
+
+# Remove blank lines from config file
+with open('tmp.txt', 'w') as file:
+    file.write( run(['grep', '-v', '-e', '^[[:space:]]*\$', "${config}"], capture_output=True, text=True).stdout )
+run(['mv', 'tmp.txt', "${config}"])
```

Comment:
I'd use `run` only with `check=True`. But in this case it would probably be easier/more elegant to just build the entire file in python and only write the final file to disk.

Reply:
Hi @grst ,
I have changed to parse all the multiline string first and then save the file just once.
If you can take a look to see if it is resolved.